### PR TITLE
[bugfix](segmentload) should remove segment from segment cache if load segment failed

### PR DESCRIPTION
**Pull request by yiguolei**
## Proposed changes

1. If the segment load failed, it is not recoverable, should remove it from segment cache and the query will create new segment again.
2. Could not depend on remove segment from cache during load index failed, because the call once may throw exception and the error handle logic is not reached.

### CODE REVIEW

**Comment by yiguolei** [comment]
run buildall

**Comment by yiguolei** [comment]
run buildall

File: be/src/olap/rowset/segment_v2/segment.cpp [hunk 1]
```diff
@@ -910,8 +916,14 @@ Status Segment::new_inverted_index_iterator(const TabletColumn& tablet_column,
     ColumnReader* reader = _get_column_reader(tablet_column);
     if (reader != nullptr && index_meta) {
         if (_inverted_index_file_reader == nullptr) {
-            RETURN_IF_ERROR(
-                    _inverted_index_file_reader_open.call([&] { return _open_inverted_index(); }));
+            auto status = [this]() {
+                return _inverted_index_file_reader_open.call(
+                        [&] { return _open_inverted_index(); });
+            }();
```
Comment (gavinchou) [inline be/src/olap/rowset/segment_v2/segment.cpp:922]: why not just 
```suggestion
            auto status = _inverted_index_file_reader_open.call([&] { return _open_inverted_index(); });
```

File: be/src/olap/rowset/segment_v2/segment.cpp [hunk 2]
```diff
@@ -554,11 +554,17 @@ vectorized::DataTypePtr Segment::get_data_type_of(const ColumnIdentifier& identi
 }
 
 Status Segment::_create_column_readers_once() {
-    return _create_column_readers_once_call.call([&] {
-        DCHECK(_footer_pb);
-        Defer defer([&]() { _footer_pb.reset(); });
-        return _create_column_readers(*_footer_pb);
-    });
+    auto status = [this]() {
```
Comment (gavinchou) [inline be/src/olap/rowset/segment_v2/segment.cpp:557]: why not just
```
auto status = _create_column_readers_once_call.call(...);
```

**Comment by yiguolei** [comment]
run buildall

**Comment by yiguolei** [comment]
run buildall

**Review by eldenmoon — APPROVED** [review]
LGTM

File: be/src/olap/segment_loader.cpp [hunk 3]
```diff
@@ -59,8 +59,14 @@ Status SegmentLoader::load_segments(const BetaRowsetSharedPtr& rowset,
     for (int64_t i = 0; i < rowset->num_segments(); i++) {
         SegmentCache::CacheKey cache_key(rowset->rowset_id(), i);
         if (_segment_cache->lookup(cache_key, cache_handle)) {
-            continue;
+            // Has to check the segment status here, because the segment in cache may has something wrong during
```
Comment (xiaokang) [inline be/src/olap/segment_loader.cpp:62]: Why not check status in lookup and erase not healthy segment?

Comment (yiguolei) [reply be/src/olap/segment_loader.cpp:62]: because the segment cache code is too complicated, it may produce memory leak issue during lookup.

**Review by xiaokang — APPROVED** [review]
LGTM

**Pull request merged**